Comorphisms/SuleCFOL2TPTP.hs

```diff
@@ -0,0 +1,682 @@
+    --                  -> Result FOF_unitary_formula
+    toUnitaryFormula x = case x of
+      Quantification q vars f _ -> do
+        let fofVars = translateVarDecls vars
+        let variableList = map fst fofVars
+        let variableDeclaration =
+              FOFUF_logic $ FOFLF_binary $ FOFBF_assoc $ FOFBA_and $
+              map (uncurry sortOfX) fofVars
+        fofF <- toUnitaryFormula f
+        case q of
+          Universal ->
+            let implication =
+                  FOFUF_logic $ FOFLF_binary $ FOFBF_nonassoc $
+                  FOF_binary_nonassoc TAS.Implication variableDeclaration fofF
+            in  return $ FOFUF_quantified $
+                         FOF_quantified_formula ForAll variableList implication
+          Existential ->
+            let conjunction =
+                  FOFUF_logic $ FOFLF_binary $ FOFBF_assoc $
+                  FOFBA_and [variableDeclaration, fofF]
+            in  return $ FOFUF_quantified $
+                         FOF_quantified_formula Exists variableList conjunction
+          -- Has been resolved/removed by prepareNamedFormula:
+          Unique_existential ->
+            fail "Unique_existential occurred where it cannot occur. This is a bug in Hets."
+      Junction Con fs _ -> do
+        fofs <- mapM toUnitaryFormula fs
+        return $ FOFUF_logic $ FOFLF_binary $ FOFBF_assoc $ FOFBA_and fofs
+      Junction Dis fs _ -> do
+        fofs <- mapM toUnitaryFormula fs
+        return $ FOFUF_logic $ FOFLF_binary $ FOFBF_assoc $ FOFBA_or fofs
+      Relation f1 CAS.Implication f2 _ -> do
+        fof1 <- toUnitaryFormula f1
+        fof2 <- toUnitaryFormula f2
+        return $ FOFUF_logic $ FOFLF_binary $ FOFBF_nonassoc $
+                 FOF_binary_nonassoc TAS.Implication fof1 fof2
+      -- For some reason, "f2 if f1" is saved as "Relation f1 RevImpl f2 _"
+      Relation f1 RevImpl f2 _ -> do
+        fof1 <- toUnitaryFormula f1
+        fof2 <- toUnitaryFormula f2
+        -- Flip the formulae:
+        return $ FOFUF_logic $ FOFLF_binary $ FOFBF_nonassoc $
+                 FOF_binary_nonassoc ReverseImplication fof2 fof1
+      Relation f1 CAS.Equivalence f2 _ -> do
+        fof1 <- toUnitaryFormula f1
+        fof2 <- toUnitaryFormula f2
+        return $ FOFUF_logic $ FOFLF_binary $ FOFBF_nonassoc $
+                 FOF_binary_nonassoc TAS.Equivalence fof1 fof2
+      Negation f _ -> do
+        fofF <- toUnitaryFormula f
+        return $ FOFUF_unary $ FOFUF_connective NOT fofF
+      Atom True _ ->
+        return $ FOFUF_atomic $ FOFAT_defined $ FOFDAF_plain $
+                 FOFDPF_proposition $ TPTP_true
+      Atom False _ ->
+        return $ FOFUF_atomic $ FOFAT_defined $ FOFDAF_plain $
+                 FOFDPF_proposition $ TPTP_true
+      Predication predSymb terms _ -> do
+        predName <- case predSymb of
+              Pred_name _ ->
+                fail "An unqualified predicate has been detected. This is a bug in Hets."
```

Comment:
in general, the message in such `fail` statements should start with the name of the module. This eases debugging.